This adds thaw.js

freddyC opened the pull request:
![A GIF to give the reviewer some spice of the internet](https://vignette.wikia.nocookie.net/tb-org-channel/images/0/04/Thaw.gif/revision/latest?cb=20161103145232)

should solve https://github.com/BrainJS/brain.js/issues/82

## Makes training async via thaw.js


### Removes `reinforce` option
I felt this cleans up and simplifies the network without changing functionality.

Basically when you train if it doesn't have a set size for the network it initializes and uses your options (as normal) to dictate network size, layers and all

If you have already trained the network or if the network is loaded from JSON then you don't need to set it and it trains as normal.

### Potential PROBLEMS
1) The train options `training callback called with training stats` test fails
    a) I think this is a problem in thaw.js but haven't been able to track it down yet.

2) Should probably have more tests for the new train options

3) With training being async now this may create a new problem with trying to train when currently training.  I haven't tested anything on that I was just thinking and about ramifications of this change.  We might want to add some sort of (is training) variable and if someone tries to train while it's currently training either push this training to the end of current training or escape out with an error or something.

freddyCfit committed:
This adds thaw.js (one failing test)

### Makes training async via thaw.js

### Removes `reinforce` option
I felt this cleans up and simplifies the network without changing functionality.

Basically when you train if it doesn't have a set size for the network it initializes and uses your options (as normal) to dictate network size, layers and all

If you have already trained the network or if the network is loaded from JSON then you don't need to set it and it trains as normal.

### PROBLEMS
1) The train options `training callback called with training stats` test fails
  - I think this is a problem in thaw.js but haven't been able to track it down yet.

2) Should probably have more tests for the new train options

3) With training being async now this may create a new problem with trying to train when currently training.  I haven't tested anything on that I was just thinking and about ramifications of this change.  We might want to add some sort of (is training) variable and if someone tries to train while it's currently training either push this training to the end of current training or escape out with an error or something.

freddyC commented on .gitignore:
```diff
@@ -6,4 +6,6 @@ node_modules
\ No newline at end of file
+yarn.lock
+
+/dist**
```
yeah I shouldn't have done this

robertleeplummerjr replied:
Can you remove the dist

freddyCfit committed:
updates training to train async or not and fixes the tests to conform

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -240,47 +255,101 @@ export default class NeuralNetwork {
+   * @param cb
+   * @returns {{error: number, iterations: number}}
+   */
+  trainAsync (data, _options = {}, cb = function() {}) {
```
Can we just use a `Promise` here?

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -240,47 +255,101 @@ export default class NeuralNetwork {
     };
+
+    if (this.sizes === null) {
+      this.initialize (data);
```
The space after a method I've not seen in javascript libraries, can we aim for `justThis()`

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -240,47 +255,101 @@ export default class NeuralNetwork {
+   * @returns {{error: number, iterations: number}}
+   */
+  trainAsync (data, _options = {}, cb = function() {}) {
+    if (typeof _options === "function") {
```
We've been in the practice of using single quotes, would you be ok migrating to such?

freddyC replied:
absolutely

robertleeplummerjr commented on test/base/likely.js:
```diff
@@ -2,7 +2,7 @@ import assert from 'assert';
 import likely from '../../dist/likely';
 
-describe('likely', () => {
+describe('likely', function () {
```
Lets keep the newer `() => {}` syntax where possible.

freddyC replied:
One thing I found out is that mocha doesn't play well with the new `() => {` syntax.  I like it better but I was getting some really weird crashes that I think had to do with that. That being said I probably went overboard in changing it to `function () {`

robertleeplummerjr replied:
Just use `--compilers js:babel-core/register` with mocha

robertleeplummerjr reviewed (changes requested):
Good job, had a few items to review.  Ty ty!

freddyCfit committed:
Rebased onto master and small update for initialize to not send data but sizes

perkyguy commented on src/neural-network.js:
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
+    let errorThresh = options.errorThresh;
+    let log = options.log === true ? console.log : options.log;
+    let logPeriod = options.logPeriod;
+    let learningRate = _options.learningRate || this.learningRate || options.learningRate;
```
Why the triple places this could exist?

freddyC replied:
The only thing I can think of is that it gives preference to as so:
1) the passed in learning rate
2) the class's set learning rate
3) the default learning rate.

It is a little odd to me but I figured it was outside the scope of this pr

perkyguy commented on src/neural-network.js:
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
+          sum += this.trainPattern (data[i].input, data[i].output, learningRate);
+        }
+
+        res.error = sum / data.length;
```
Why not just `res.error = data.reduce(...)/data.length`?

freddyC replied:
I was talking to @robertleeplummerjr about it yesterday and we talked about the cost of closures.  They aren't crazy expensive but there is a cost to them and I don't usually take them into consideration.  I am open to either way but where it was simple to make the for loop and iterate I did that.

So I am open to either way :)

perkyguy commented on src/neural-network.js:
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
+
+        if (callback && (res.iterations % callbackPeriod === 0)) {
+          // JSON.parse/stringify to clone the object so the callback doesn't have side effects to training
+          callback(JSON.parse(JSON.stringify(res)));
```
`Object.assign` would be a little clearer for this?

freddyC replied:
👍  I didn't think of that but I will get it changed

perkyguy commented on test/base/bitwise.js:
```diff
@@ -5,46 +5,88 @@ let wiggle = 0.1;
+  data.forEach(d => {
+    var actual = net.run(d.input)
+    var expected = d.output;
+    assert.ok(actual < (expected + wiggle) && actual < (expected + wiggle), `failed to train "${op}" - expected: ${expected}, actual: ${actual}`);
```
Copy-pasta error? You have the same thing tested in this &&

perkyguy replied:
Why not just do a `.toFixed(1)`?

freddyC replied:
Good idea on the `.toFixed (1)` and yeah that was meant to clamp that value.

perkyguy commented on test/base/bitwise.js:
```diff
@@ -5,46 +5,88 @@ let wiggle = 0.1;
+    data.forEach(d => {
+      var actual = net.run(d.input)
+      var expected = d.output;
+      assert.ok(actual < (expected + wiggle) && actual < (expected + wiggle), `failed to train "${op}" - expected: ${expected}, actual: ${actual}`);
```
Same thing tested?

perkyguy reviewed (commented):
What's the use case for the training time?

perkyguy commented on src/neural-network.js:
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
+
+      if (callback && (res.iterations % callbackPeriod === 0)) {
+        // JSON.parse/stringify to clone the object so the callback doesn't have side effects to training
+        callback (JSON.parse (JSON.stringify (res)));
```
`Object.assign`?

robertleeplummerjr replied:
or `{...res}`, which is essentially the same thing.

perkyguy commented on src/neural-network.js:
```diff
@@ -223,13 +223,36 @@ export default class NeuralNetwork {
+    let outputSize = data[0].output.length;
+    let hiddenSizes = this.hiddenSizes;
+    if (!hiddenSizes) {
+      sizes.push(Math.max(3, Math.floor(inputSize / 2)));
```
If someone doesn't want hidden layers, the library will force it?

freddyC replied:
yep, that's how it is currently setup.

perkyguy replied:
Why? Can @robertleeplummerjr she's some light on this?

perkyguy replied:
Sorry, I read this as "yeah I did it this way, what of it?" Not as you probably intended it, "maintaining current functionality"

freddyC replied:
haha no worries, it's finding the balance between not changing too much but still changing things in a good way :)

robertleeplummerjr replied:
Thats how it was when we got here, however one could do: `{ hiddenSizes: [] }`

perkyguy commented on src/neural-network.js:
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
-      if (callback && (i % callbackPeriod === 0)) {
-        callback({ error: error, iterations: i });
+
+      if (callback && (res.iterations % callbackPeriod === 0)) {
```
The callback is for getting periodic training results? Not the finished trained model?

freddyC replied:
yeah that's how it is currently setup.

perkyguy replied:
Why?

robertleeplummerjr replied:
logging for the most part.  Most people like to see training advance.  I don't see this going away any time soon, as it was part of the model when the community took over.

perkyguy replied:
Isn't that what the `log` option does? It's cool though, separate conversation for the community, this maintains current functionality

perkyguy commented on src/neural-network.js:
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
+          log(`iterations: ${res.iterations} training error: ${res.error}`);
+        }
+
+        if (callback && (res.iterations % callbackPeriod === 0)) {
```
Like above, the callback is for partially trained nets?

robertleeplummerjr replied:
that is correct as well

perkyguy commented on src/neural-network.js:
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
+    const items = new Array(iterations);
+    const thaw  = new Thaw (items, {
+      delay: true,
+      each: () => {
```
This is the same as the standard `train` behavior, just delivered per iteration. This should be pulled out into its own function/method to keep it DRY

freddyC replied:
👍

robertleeplummerjr replied:
I agree, that'd be more ideal.

freddyCfit committed:
Committing before I change to promise so I don't break things too bad

freddyCfit committed:
Updates all the things.

Converts the async method to a promise, not a cb

Updates tests

Helps code be more dry

freddyCfit committed:
Last things

Adds back the () => in tests

Ran `npm run make` to get things

freddyCfit committed:
Loggity

Simplified log test, and added in the ability to pass in custom log method.

I thought this was outside of the scope of my pr but just kinda did it because the test output bugged me lol.

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -63,7 +63,7 @@ export default class NeuralNetwork {
    * @param {Number[]} sizes
    */
-  initialize(sizes) {
+  initialize (sizes) {
```
unneeded space

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -223,13 +223,36 @@ export default class NeuralNetwork {
    * @returns {{error: number, iterations: number}}
    */
-  train(data, _options = {}) {
+  train (data, _options = {}) {
```
unneeded space

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
+   * @param data
+   * @returns sizes
+   */
+  _getSizesFromData (data) {
```
uneeded space

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
+   * if false passed in nothing is logged
+   * @returns error
+   */
+  _setLogMethod (log) {
```
unneeded space

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
+   * @param learning Rate
+   * @returns error
+   */
+  _calculateTrainingError (data, learningRate) {
```
unneeded space

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
+  _calculateTrainingError (data, learningRate) {
+    let sum = 0;
+    for (let i = 0; i < data.length; ++i) {
+      sum += this.trainPattern (data[i].input, data[i].output, learningRate);
```
unneeded space

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
+   * @param status { iterations: number, error: number}
+   * @param options
+   */
+  _checkTrainingTick (data, status, options) {
```
unneeded space

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
    */
-  train(data, _options = {}) {
-    const options = Object.assign({}, this.constructor.trainDefaults, _options);
+  train (data, _options = {}) {
```
unneeded space

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
-  train(data, _options = {}) {
-    const options = Object.assign({}, this.constructor.trainDefaults, _options);
+  train (data, _options = {}) {
+    let options = Object.assign({}, this.constructor.trainDefaults, _options);
```
`options` is never reassigned, use a `const`, not a `let`.

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -22,7 +22,7 @@ export default class NeuralNetwork {
       callback: null,
       callbackPeriod: 10,
-      reinforce: false
+      trainTimeMs: -Infinity
```
Can the `trainTimeMs` be put into a seperate pr?  Want to keep the pr's as lean as possible.

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
+   * @param cb
+   * @returns {{error: number, iterations: number}}
+   */
+  trainAsync (data, _options = {}) {
```
unneeded space, this line and next

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
+      };
+
+      if (this.sizes === null) {
+        let sizes = this._getSizesFromData (data);
```
unneeded space, this line and next.

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
-      }
-    }
+      const items = new Array(options.iterations);
+      const thaw  = new Thaw (items, {
```
unneeded space

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
+      const thaw  = new Thaw (items, {
+        delay: true,
+        each: () => {
+          this._checkTrainingTick (data, status, options);
```
unneeded space

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -288,7 +370,7 @@ export default class NeuralNetwork {
    * @param learningRate
    */
-  trainPattern(input, target, learningRate) {
+  trainPattern (input, target, learningRate) {
```
unneeded space

robertleeplummerjr commented on src/neural-network.js:
```diff
@@ -634,8 +716,8 @@ export default class NeuralNetwork {
    */
-  fromJSON(json) {
-    this.initialize(json.sizes);
+  fromJSON (json) {
```
unneeded space

robertleeplummerjr replied:
still one space 😄

robertleeplummerjr commented on test/base/bitwise.js:
```diff
@@ -3,48 +3,102 @@ import brain from '../../src';
 
 let wiggle = 0.1;
 
+function isAround (actual, expected) {
```
unneeded space

robertleeplummerjr commented on test/base/bitwise.js:
```diff
@@ -3,48 +3,102 @@ import brain from '../../src';
 }
 
-describe('bitwise functions', () => {
+function testBitwiseAsync (data, op, done) {
```
unneeded space

robertleeplummerjr commented on test/base/bitwise.js:
```diff
@@ -3,48 +3,102 @@ import brain from '../../src';
+function testBitwiseAsync (data, op, done) {
+  let net = new brain.NeuralNetwork();
+  net
+    .trainAsync (data, { errorThresh: 0.003 })
```
there are several more of these, can you just handle all of them?

freddyCfit committed:
Remember that time

When my friend wanted me to give him a Charmander tattoo and he got this https://timedotcom.files.wordpress.com/2015/04/screen-shot-2015-04-17-at-11-12-39-am.png

freddyCfit committed:
Removed goodies

`trainTimeMs` is gone and will be in a new pr

`log` doesn't accept functions but will be in a new pr

The tests pass (ugly but they pass lol)

https://memegenerator.net/img/instances/75216624.jpg

good thing no one actually looks at commit messages amirite?

robertleeplummerjr commented on test/base/hash.js:
```diff
@@ -2,79 +2,174 @@ import assert from 'assert';
 
 describe('hash input and output', () => {
-  it('runs correctly with array input and output', () => {
+  it ('runs correctly with array input and output', () => {
```
(space)

robertleeplummerjr commented on test/base/likely.js:
```diff
@@ -62,23 +66,24 @@ describe('likely', () => {
-      '#.....#' +
-      '###.##.'
-    ), net);
+      let result = likely(character(
```
spacing is odd here

robertleeplummerjr reviewed (approved):
This looks great.  I had a couple suggestions, but they don't change functionality, it was just spacing.  Ty for your hard work!

robertleeplummerjr committed:
Merge branch 'develop' into add-thaw

# Conflicts:
#	test/base/bitwise.js
#	test/base/json.js
